### PR TITLE
Improve GC

### PR DESCRIPTION
**Pull request by zvin**
---- Autogenerated Waffleboard Connection: Connects to #17

### CODE REVIEW

File: lib/docker-image-tree.coffee [hunk 1]
```diff
@@ -1,13 +1,21 @@
...
 	return if '<none>:<none>' in repoTags then [] else repoTags
 
 exports.createNode = createNode = (id) -> { id: id, size: 0, repoTags: [], mtime: null, children: {} }
 
-exports.createTree = createTree = (images) ->
+getMtime = (tree, layer_mtimes) ->
+	mtime = layer_mtimes[tree.id]
+	if mtime == undefined
+		key = _.head(_.intersection(_.keys(layer_mtimes), tree.repoTags))
+		if key != undefined
```
Comment (CameronDiver) [inline lib/docker-image-tree.coffee]: `if key?`

Comment (zvin) [reply lib/docker-image-tree.coffee]: Updated

**Comment by zvin** [comment]
>This looks good I guess, but wouldn’t just trying to remove all images and keeping a list of those that failed in order to retry somewhat later be as effective but simpler strategy?

@dfunckt When a removal fails, in general it's because the image has a new child image or a new container using it. It will most probably fail again.
Restarting the garbage collection will estimate the required space and get the tree of images again.

For example, consider this image tree:
```
     G
    / \    
   /   \
  E     F
 / \   / \
A   B C   D
           \
            H
```
 * no image is being used by a container
 * the oldest and larges images are, in order, H, D, C and F and their size is enough for `reclaimSpace`

The GC will select to remove H, then D then C then F.
If removing H fails, all following removals will fail (except for C).

**Comment by dfunckt** [comment]
Yeah that makes sense, given we allow a specific "reclaimSpace" to be specified.

Consider how much simpler this library would be if we always tried to remove everything (without force) and the "retry" list was retried in random order. That way, at some point, every leaf image would be removed, allowing parent images to be removed as well.

Also, I suspect we care about the module being general enough, but what's *our* use case to care about containers? As far as I know, we only use this library in the builder which never starts any containers.

Anyway, it's not so important (and I might be missing something) and given you've already fixed the library to work, it's good enough for me.

**Comment by zvin** [comment]
@dfunckt 
We care about containers because there some builds or a `df` (used to estimate if we need to GC) running and removing images they use will fail.

**Pull request merged**